### DISCUSSION TITLE
What is this data structure? (b'zzzz', ('192.168.4.16', 43578))

### DISCUSSION THREAD

**Discussion by boylesg** [micropython/micropython#9629] [Using MicroPython]
Obtained via this code

ValuePair = self.Socket.recvfrom(self.nReceiveDataSize)
print(ValuePair)

What is the 'b' about?

'zzzz' is the data I sent from my mobile phone over Wifi access point.

It looks to me like an array of 2 elements where the first element is the string 'zzzz' and the second element is another array.

**Comment by jimmo** [comment]
It's a bytes string (i.e. a string of bytes). https://docs.python.org/3/library/stdtypes.html#bytes

This is different to a regular string which is a string of characters (which may not actually be bytes, e.g. utf-8 encoded).

You can use ValuePair.decode() to decode bytes into a string.

**Reply by jimmo** [reply]
Note: this is standard Python behaviour -- nothing Micropython specific here.

**Reply by boylesg** [reply]
> It's a bytes string (i.e. a string of bytes). https://docs.python.org/3/library/stdtypes.html#bytes
> 
> This is different to a regular string which is a string of characters (which may not actually be bytes, e.g. utf-8 encoded).
> 
> You can use ValuePair.decode() to decode bytes into a string.

I am still not sure I entirely understand what it is.

It looks like an array within an array.
First element is a string 'zzzz'
Second element is another array containing IP address and port.
Is that accurate in anyway?

My coding instinct is that [0] is 'zzzzzz'
And [1] is the other arrray containing ip address and port.

[1][0] is ip address
[1][1] is port
Correct?

Essentially assuming I have this:

B = (b'zzzz', ('192.168.4.16', 43578))

How do I convert B to 3 separate strings containing 'zzzzzz', '192.168.4.16' and '43578' (or just an integer 43578)

From the link you posted it looks as though 'b' is  a separator value. 
But what does 'b' mean exactly.
Is 'zzzzzz' seperated from '192.168.16' literally via 'b'?
Or they separated by the integer/ASCII value contained in a struct field named 'b'?
In which case B.b would give me the separator value.
But then what field of stuct B gives me 'zzzz', ('192.168.4.16', 43578)?

Very confusing!

I just don't have a mental diagram of what this data structure is.

**Reply by jimmo** [reply]
the return from revcfrom is a tuple containing a bytes string and another tuple of string and int.

yes, a tuple is like a list (it's a standard python data type).

the b before the quotes tells you it's a byte string. b"foo" is a bytes string, but "foo" is a regular string.

b"\xe2\xad\x90" is the same underlying data as "⭐"

```
data = socket.recvfrom(...)
message = data[0].decode()
ip, port = data[1]
```

that last line us using unpacking assignment, it's the same as writing
```
ip = data[1][0]
port = data[1][1]
```

**Reply by boylesg** [reply]
> 

Should I be calling decode on these as well?

ip = data[1][0]
port = data[1][1]

**Reply by jimmo** [reply]
no, ip is a string (it doesn't have the b prefix) and port is an integer.

**Reply by boylesg** [reply]
> no, ip is a string (it doesn't have the b prefix) and port is an integer.

Ahhh OK.

**Reply by jimmo** [reply]
the point is that the network is transmitting raw data (i.e. bytes) so that's how python gives it to you. Whereas the ip and the port make sense to be a specific type (string and int, respectively)

**Reply by boylesg** [reply]
> the point is that the network is transmitting raw data (i.e. bytes) so that's how python gives it to you. Whereas the ip and the port make sense to be a specific type (string and int, respectively)

I am getting my head around it.

Just to add more complication to all this.....

To stop function recvfrom etc. from blocking until data is received you have to set a timeout on the socket.

But if you do that then recvfrom  throws a friggin exception when there is no data incoming.

So you have to do this sort of thing to stop that crashing your sketch.
                try:
                    ReturnValue = self.Client.recvfrom(self.nReceiveDataSize)
                except OSError:
                    strRequest = ""

What a friggin mess!!!!

I thought Python was supposed to be the bees knees in being easier than C.

Friggn NOT - it is just as complicated and messy as C, if not worse. LOL

The syntax error messages are certainly worse than C.

At least with C you will get syntax error messages along the lines of "; expected" 

Python syntax errors give you no more than "syntax error" leaving you scratching your head for a bit because there is a missing :

**Reply by robert-hh** [reply]
The nonblocking aspect is standard socket behavior. If  you tell the code to block it will do so.

The error handling in Python is different that that of C. So using a try: / except: structure is the Python style of dealing with errors, instead of the C style of checking return codes. It may look clumsy at first glance, but has advantages. Different to C you can put a whole sequence of code into a try: section, and catch any error that may happen in the sequence, instead of checking each and every action that may fail. 

But I agree that the syntax error messages are quite terse in MicroPython, The implementation is meant to be compact. The standard CPython is way more informative about Syntax error and points at the place which is wrong. So please do not mix up the MicroPython implementation with general aspects of the Python language.

**Reply by boylesg** [reply]
> The nonblocking aspect is standard socket behavior. If you tell the code to block it will do so.
> 
> The error handling in Python is different that that of C. So using a try: / except: structure is the Python style of dealing with errors, instead of the C style of checking return codes. It may look clumsy at first glance, but has advantages. Different to C you can put a whole sequence of code into a try: section, and catch any error that may happen in the sequence, instead of checking each and every action that may fail.
> 
> But I agree that the syntax error messages are quite terse in MicroPython, The implementation is meant to be compact. The standard CPython is way more informative about Syntax error and points at the place which is wrong. So please do not mix up the MicroPython implementation with general aspects of the Python language.

Well C++ has try/catch mechanism as well. But it has probably been 'retrofitted' to C++, and not all libraries use it.


I

**Reply by boylesg** [reply]
> The nonblocking aspect is standard socket behavior. If you tell the code to block it will do so.
> 
> The error handling in Python is different that that of C. So using a try: / except: structure is the Python style of dealing with errors, instead of the C style of checking return codes. It may look clumsy at first glance, but has advantages. Different to C you can put a whole sequence of code into a try: section, and catch any error that may happen in the sequence, instead of checking each and every action that may fail.
> 
> But I agree that the syntax error messages are quite terse in MicroPython, The implementation is meant to be compact. The standard CPython is way more informative about Syntax error and points at the place which is wrong. So please do not mix up the MicroPython implementation with general aspects of the Python language.

I am trying to put all this in a library for our coder dojo kids so that we are not debugging multiple versions of this same code for dozens of projects.

A simple function call to start a UDP server, HTTP server, AP UDP server or AP HTTP server

And then a single function for polling for requests no matter which of the above is chosen.

And a single function to send a response to the 'requester'

Next is to get my head around how to send a web page for HTTP server.
It appears as though it is not as simple of just sending a plain old string containing your html :-(

**Reply by jimmo** [reply]
Non-blocking sockets raise exceptions if you read and there's no data because you're not supposed to do that.

The idea is that you're supposed to have figured out if there's data e.g. using select() or poll(). This is exactly the same as in C, where you'd be using select/poll/epoll/kqueue.

In Python, the select/poll thing is simplified dramatically using asyncio, however I'm not sure if Micropython's asyncio supports UDP fully yet.

As an exercise, consider how you'd design a version of recvfrom that didn't raise an exception in this situation. Returning None is no better because you still have to handle it.

**Reply by boylesg** [reply]
> Non-blocking sockets raise exceptions if you read and there's no data because you're not supposed to do that.
> 
> The idea is that you're supposed to have figured out if there's data e.g. using select() or poll(). This is exactly the same as in C, where you'd be using select/poll/epoll/kqueue.
> 
> In Python, the select/poll thing is simplified dramatically using asyncio, however I'm not sure if Micropython's asyncio supports UDP fully yet.
> 
> As an exercise, consider how you'd design a version of recvfrom that didn't raise an exception in this situation. Returning None is no better because you still have to handle it.

You HAVE to be able to poll data sources without them necessarily blocking!

Consider an RC control robot thing that does nothing else but sit there as wait for commands to move it is one direction or another - blocking is fine,

But consider a weather station that logs weather data to an SD card but also services HTTP requests to view the data to date.

If the socket blocks until a HTTP request is received then no weather data will ever be logged!!!!!!

**Reply by cwalther** [reply]
> You HAVE to be able to poll data sources without them necessarily blocking!

That is correct, and nothing of what jimmo wrote contradicts it. Read https://docs.micropython.org/en/latest/library/select.html to find out how.

**Reply by boylesg** [reply]
> > You HAVE to be able to poll data sources without them necessarily blocking!
> 
> That is correct, and nothing of what jimmo wrote contradicts it. Read https://docs.micropython.org/en/latest/library/select.html to find out how.

At least for MCU, programs where debugging tools tend to be rather rudimentary, I am really not a fan of introducing yet more layers of complexity with this 'poll' class.

It starts making things less simple and less convenient to debug.

I really prefer to just have direct control over how long those receive functions block for.

As far as the sort of asynchronous projects I am working on, not returning any data after a poll is not an error condition.

Threads (effectively) might be OK for regular python where you probably have access to proper integrated debuggers in which you can set break points etc.

But I think they should be used vary sparingly (if at all) in MCU projects, where you are stuck with the likes of Thonny.

**Reply by cwalther** [reply]
For what it’s worth, empirically (on an ESP32) I find that I can use `read()` instead of `recv[from]()` to read from a non-blocking UDP socket, and it will return `None` instead of raising an exception when there is no data available. I don’t see that specified anywhere though (in CPython, `makefile()`, which gets one the `read` method, is explicitly specified to be unsupported on non-blocking sockets). Checking [the](https://github.com/micropython/micropython/blob/master/extmod/modusocket.c) [source](https://github.com/micropython/micropython/blob/master/py/stream.c), it depends on whether the underlying network stack returns `EAGAIN/EWOULDBLOCK` or some other error.
